src/main/java/com/launchdarkly/client/FeatureFlag.java

```diff
@@ -67,40 +67,39 @@ EvalResult evaluate(LDUser user, FeatureStore featureStore) {
     }
     List<FeatureRequestEvent> prereqEvents = new ArrayList<>();
     Set<String> visited = new HashSet<>();
-    return evaluate(user, featureStore, prereqEvents, visited);
+    JsonElement value = evaluate(user, featureStore, prereqEvents, visited);
+    return new EvalResult(value, prereqEvents);
   }
 
-  // Returning either a nil EvalResult or EvalResult.value indicates prereq failure/error.
-  private EvalResult evaluate(LDUser user, FeatureStore featureStore, List<FeatureRequestEvent> events, Set<String> visited) {
+  // Returning either a JsonElement or null indicating prereq failure/error.
+  private JsonElement evaluate(LDUser user, FeatureStore featureStore, List<FeatureRequestEvent> events, Set<String> visited) {
     boolean prereqOk = true;
-    EvalResult evalResult = new EvalResult(null, events, visited);
     for (Prerequisite prereq : prerequisites) {
-      evalResult.visitedFeatureKeys.add(key);
-      if (evalResult.visitedFeatureKeys.contains(prereq.getKey())) {
+      visited.add(key);
```

Comment:
This line is in the wrong place. You're adding the feature flag's key with each iteration over the prerequisites.